### PR TITLE
[WIP] House search

### LINKED ISSUE #25: Integration with search engines
Searching will be critical component for Mağara because nobody wants to check
every single page to find a house.

I have played with Elasticsearch 5, but it doesn't mean that it is only one in
the world<sup>[1]</sup>, so we should consider other search engine options such
as, Solr, Algolia etc.

## What will we achieve with a search engine?

Yeah, we need to ask before diving into code, right? Well, like I said before
*"searching will be our backbone"*.

* Searching with in the map (latitude and longitude are ready for this). Being
capable to search with advanced features like Checkbox (model), rent or rating
* Analyzing logs to find problems - further versions, not now

## Why Elasticsearch?

I prefer Elasticsearch as a sysadmin or a Rails developer since I can use for
searching houses or logs (with logstash), and also, it provides Kibana, ES
dashboard, to visualize data that we have.

We have different gems in Rails community like:

* [`elasticsearch-rails`][elasticsearch-rails] and `elasticsearch-model`:
  Offical Rails gem to communicate ES API.
* [`chewy`][chewy]: High level framework which is based on `elasticsearch-ruby`
* [`searchkick`][searchkick]: *Batteries included* search solution

[1]: https://db-engines.com/en/ranking/search+engine
[elasticsearch-rails]: https://github.com/elastic/elasticsearch-rails
[chewy]: https://github.com/toptal/chewy
[searchkick]: https://github.com/ankane/searchkick

### PR DESCRIPTION
**Pull request by noloman**
Fixes #25 **DO NOT MERGE YET**

I have created a minimal search for houses, based on several fields of the houses and address table.
I still need to add some RSpec specs and check if there's a better way of doing things.
Please check all out and let me know if there's improvements to be done and if it's all correct.
Thanks!

### CODE REVIEW

**Comment by noloman** [comment]
Hey @BerkhanBerkdemir, yes I'm trying first to build a small search engine (right now, I'm struggling a bit with RSpec and its specs for this PR). 
I decided not to go for ElasticSearch for now since it's only going to be a small search functionality here, but I guess in the future, if we want a broader search feature, we should go for ElasticSearch.

Let me know what you think! And good luck with college!

File: app/models/house.rb [hunk 1]
```diff
@@ -23,4 +23,17 @@ def address
   def checkbox
     super || build_checkbox
   end
+
+  def self.search(term)
+    if term
+      House.joins(:address)
```
[the remark this answers is not in the document]
Comment (noloman) [reply app/models/house.rb]: yes, that's a problem, but I've updated the PR

**Comment by noloman** [comment]
I have replaced the ordinary search for elasticsearch using searchkick (https://github.com/ankane/searchkick). It works while searching for example in the description of the house, but for some reason, it doesn't work by searching the name of the city

**Comment by BerkhanBerkdemir** [comment]
Hey @noloman,

Here is the patch for `docker-compose.yml`. I want to add this patch because developers don't need to worry about setting up a Elastichsearch.

```
Author: Berkhan Berkdemir <24360355+BerkhanBerkdemir@users.noreply.github.com>
Date:   Wed Nov 28 18:40:18 2018 -0800

    :whale: add Elasticsearch 5 Alpine container
    
    If somebody wants to run Elasticsearch as a container, they
    can easily spin up `docker-compose.yml`. However, we should
    be careful because this container uses version 5 of
    Elasticsearch, but searchkick gem uses 6. So, we could
    have some problem because of API incompatibility.
```

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 6b33eb8..f47b1d4 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -12,5 +12,15 @@ services:
     networks:
     - backend
 
+  # XXX: Searchkick uses ES 6 API, but we provide ES 5 server because Docker Hub
+  # does not provide any lightweight container for this purpose.
+  elasticsearch:
+    image: elasticsearch:5-alpine
+    ports:
+    - "9200:9200"
+    - "9300:9300"
+    networks:
+    - backend
+
 networks:
   backend:
```

**Comment by BerkhanBerkdemir** [comment]
I checked Semaphore's documentation, and I found [that](https://semaphoreci.com/docs/elasticsearch.html). So, that says that Elasticsearch runs automatically, but `sidekiq` tries to connect `localhost:9200` and it doesn't connect to ES. Do we know how to change Elasticsearch URL?

**Comment by noloman** [comment]
> I checked Semaphore's documentation, and I found [that](https://semaphoreci.com/docs/elasticsearch.html). So, that says that Elasticsearch runs automatically, but `sidekiq` tries to connect `localhost:9200` and it doesn't connect to ES. Do we know how to change Elasticsearch URL?

I have found https://stackoverflow.com/questions/14379575/configure-port-number-of-elasticsearch, but I guess we need to configure that in semaphore, right?

**Comment by BerkhanBerkdemir** [comment]
> Hi,
>
> Thanks for the message.
>
> The Elasticsearch service is not running by default, even though the documentation page says otherwise. To better optimize the resources during the builds on the Standard platform, Elasticsearch service was disabled and left to the user to decide if it will be used or not. Unfortunately, Semaphore docs were not updated accordingly.
>
> To start the Elasticsearch service, please add the following command to the setup of your project:
> ```
> sudo service elasticsearch start
> ```
> This should allow you to connect and use it in your builds.
>
> Also, I will create the PR with updates to the documentation page.
>
> I'm sorry for the inconvenience this might have caused you.
>
> If there's anything else I could help you with, feel free to reach out.
>
> Best,
>
> --
> Goran *
> Semaphore Support
>
>
> On Thu, Nov 29, 2018 at 21:47:29 CET, Berkhan Berkdemir <> wrote:
> > Hi,
> >
> > We have some problem about Elastichsearch testing on CI side. The problem is sidekick (Ruby gem which is client for Elasticsearch) cannot find a localhost Elasticsearch service. Could you guide us about that and renew the documentation?
> >
> > Thank you.

So, I will add `sudo service elasticsearch start` in the command line list and trigger that PR again.

**Comment by BerkhanBerkdemir** [comment]
They didn't turn back to my email, @noloman :disappointed:

**Pull request closed without merging**